README.md
# 南七技校健康打卡平台模拟打卡、跨校区报备脚本

此脚本使用统一身份认证接口的打卡网页进行自动打卡、跨校区报备和进出校申请。由于某些原因，本脚本暂不提供过验证码功能。

## 注意与建议
1. 此脚本仅适用于已返校、正常在校园内的健康同学。
2. 此脚本仅用于替代网页点击行为，使用此脚本视为使用者模拟点击操作，作者不对由于使用此脚本而产生的任何后果负责。
3. 请规范化使用脚本，切勿进行非法行动。

## 脚本使用说明
1. 需要Python3环境（见[下文](#Python版本需求)）。
2. `daily-report.py`为仅打卡脚本，适用于无需出校的同学。
3. `daily-report-cross-campus.py`为打卡+跨校区申请的脚本，适用于有“在校可跨校区”权限的同学。
4. `daily-report-out-campus.py`为打卡+进出校申请的脚本，适用于有“在校可跨校区”权限的同学且在校外居住的同学。
5. 脚本默认关闭了调试输出，并开启了自动延迟以模拟更加真实的打卡行为。
6. 脚本采用`-d`参数打开调试输出，`-i`参数控制无延迟的打卡（主要用于调试）。可同时使用，例如`python3 daily-report.py -di`

## Python版本需求
请使用Python3.x版本（作者系统上安装的版本为`3.7.3`）。

运行前请务必安装`requests`和`lxml`包。安装方法（使用pip）
```
$ pip3 install requests lxml
```

## 使用方法
1. 克隆或下载本项目。
2. 请修改脚本头部必需的相关个人信息。（修改后请勿分享给他人以防信息泄漏。）
3. 运行脚本`daily-report`开头的脚本即可打卡或打卡并申请跨校区报备、进出校申请。
4. 可为此脚本创建自启动任务来自动打卡（具体方法请自行搜索），但作者不为此行为负责。

注意：脚本没有考虑密码错误的情况。

## 脚本逻辑
1. 获取自动打卡的统一身份认证系统登录页，从页面的提交表单中分离出CAS-Token键值对。
2. 对自动打卡的统一身份认证系统界面，提交用户名和密码和上面获取到的，并从多次重定向的页面中获取所有Cookies
3. 在已有Cookies的情况下会得到打卡界面。从页面的提交表单中分离出`_token`键值对。
4. 将`_token`和相应选项一并提交。
5. 申请报备的页面分为三次请求，第一个页面（跨校区/进出校/离校申请三合一）逻辑比较复杂（详见`doc/cross-campus.md`文件），但可以跳过（本项目对校内居住同学只关注跨校区报备，其它情况请手动操作！）。带报备的脚本会模拟请求一次第一个页面，然后访问第二个页面（选择出校时间和目的地）分离出`_token`键值对。
6. 带报备的脚本最后会带着`_token`和相应选项一并提交报备。
7. 申请进出校的页面同样为三次请求，但第一个页面比较简单，正常情况下只需要选择进入校内。因此直接重定向到`t=3`的第二个页面。
8. 申请进出校的页面提交时需要选择老师审核，提交的表单中需要老师编号，需要用`check-advisor-id.py`获取。请为每次申请负责！


ChangeLog.md
# ChangeLog

## V3.5: 2022-05-05 更新

1. 适配系统中新增的“进出校报备”功能（区别于原来的“进出校申请”），此功能不需要审核。

## V3.4: 2022-04-07 更新

1. 适配进出校申请时可申请后一天，无须当天申请。
2. 进出校申请默认每天只申请后一天，这样导师/班主任有充分的时间来审核。如需要申请当天，手动在系统中申请即可。

## V3.3: 2022-04-06 更新

1. 适配跨校区报备和进出校申请时由可申请当天加后一天改为只能申请当天的问题。
2. 已知未适配：晚上9点之后报备时间为24小时（非当天）。因此打卡脚本请在当天时间使用。

## V3.2: 2022-03-30 更新

1. 适配居住地为“合肥市内校外”的同学需在系统中填写当前住址。

## V3.1: 2022-03-22 更新

校外居住的同学状态为“在线”和“合肥市内校外”，特殊情况申报转换后已无法使用跨校区报备，必须使用“进出校申请”。故增加对应脚本`daily-report-out-campus.py`。

1. 增加日常打卡+进出校申请的脚本。（该脚本申请后，请联系老师通过）
2. 适配对跨校区目的地的选择。
3. 增加`check-advisor-id.py`脚本用于查看老师编号（用于自动提交时的表单数值）。

## V3.0: 2022-03-20 更新

打卡/报备系统更新，每周申请一次的出校脚本已无法使用。在打卡脚本的基础上增加跨校区报备功能的脚本。

1. 增加日常打卡+跨校区报备的脚本。
2. 申请报备界面新增了“跨校区报备”功能，并和之前“进出校申请”、“离校申请”合并为“出校申请报备”，且此页面逻辑复杂。本项目仅支持跨校区报备功能，只对其中两种情况（高新校区到先研院往返、老校区间往返）进行处理。其他情况均为“进出校申请”和“离校申请”，请自行在网页上提交（需要老师审核）。
3. 适配对跨校区目的地的选择。

## V2.2: 2022-03-19 更新

1. 打卡界面取消了当前位置（省、市、区）的选项，已从脚本中移除。
2. 打卡界面的当前居住地和校区不再以数字2-0编号，而直接以文字形式提交校区信息，已作适配。

## V2.0: 2021-08-22 更新
统一身份认证系统逻辑已重做，打卡界面也已多次修改，并增加紧急联系人情况，原脚本已不能使用。
1. 按统一身份认证系统的新逻辑，首先获取登录界面和CAS token，再输入用户名密码登录得到打卡界面的token。
2. 增加对紧急联系人信息的支持。
3. 减少冗余代码并重构，将三次访问过程封装为三个函数。
4. 取消两个打卡脚本用于延时的设置，增加-s和-i选项用于控制输出和关闭自动延迟。


daily-report.py
#!/usr/bin/env python3

######################################################
# 自动打卡的参数，请在这改

## 必填信息：

# 科大统一认证账号密码
# （明文保存。请勿分享带密码的脚本给他人）
USERNAME = ''
PASSWORD = ''

# # 当前所在地：省、市、区
# # 具体数值和身份证前六位编码方式相同。
# # 省、市选择时分别只保留前二、四位，后面补0至六位。
# #   不知道可自行上网搜，这里提供合肥市的常用编码
# # 安徽省
# NOW_PROVINCE = '340000'
# # 合肥市
# NOW_CITY = '340100'
# # 包河区：340111，蜀山区：340104
# NOW_DISTRICT = '340111'
# #NOW_DISTRICT = '340104'

# 当前状态：
# 1: 正常在校园内
# 2: 正常在家
# 如非这两种情况请手动打卡
NOW_STATUS = '1'

# # 校区：
# # 东(2)，南(3)，中(4)，北(5)，西(6)，先研院(7)，
# # 国金院(8)，其他校区(9)，校外(0)
# WHICH_CAMPUS = '6'

# 2022/03/19更新：校区不再以数字编号，而直接以汉字作为值
# 可能的值：东（/西/南/北/中/高新）校区、先研院、国金院、合肥市内校外、合肥市外校区
CAMPUS_NAME = '西校区'

# 2022/03/30更新：校外居住报备时需要填写具体位置。校内同学可跳过
## 区县选择：只能选择以下四个区之一：包河区、蜀山区、庐阳区、瑶海区
HOME_DIST = ''
#HOME_DIST = '包河区'
HOME_ADDR = ''

# 紧急联系人：姓名、关系、电话
EMERGENCY_CONTACT = ''
EMERGENCY_CONTACT_RELATION = ''
EMERGENCY_CONTACT_PHONE = ''

# 后面的内容不需要修改。如不能用请提issue
######################################################
import sys
import requests
from requests.cookies import RequestsCookieJar
from lxml import etree

import time
import random

import argparse

# Arguments from command line
args = None

dout = lambda line: (args.debug and print(line))
sleep = lambda n: (not args.no_sleep and time.sleep(n))

def parse_args():
	global args
	parser = argparse.ArgumentParser()
	parser.add_argument('-d', '--debug', action='store_true', help='Debug mode: show debug info')
	parser.add_argument('-i', '--no-sleep', action='store_true', help='Debug mode: no sleep')
	return parser.parse_args()

def check_global_input():
	# Check username and password
	global USERNAME, PASSWORD
	if not USERNAME:	USERNAME = input('请输入学号：')
	if not PASSWORD:	PASSWORD = input('请输入密码：')

def visit(req, fn_name, url, cookie_jar, payload=None, headers=None, param=None):
	fn_list = {'get': req.get, 'post': req.post}
	try:
		fn = fn_list[fn_name.lower()]
	except KeyError:
		raise ValueError('Unsupported visit method: %s' % fn_name)

	dout('Requesting %s' % url)
	r = fn(url, cookies=cookie_jar, data=payload, headers=headers, params=param,
		allow_redirects=False, timeout=50)
	# Trace redirections: keep all cookies
	while r.status_code in range(300, 304):
		url = r.headers['Location']
		dout('Redirecting to %s' % url)
		cookie_jar.update(r.cookies)
		r = req.get(url, cookies=cookie_jar, allow_redirects=False, timeout=50)
	
	# Finally update my cookies
	cookie_jar.update(r.cookies)
	return r

def visit_login_page(req, cookie_jar):
	# Get CAS login page
	r = visit(req, 'get',
	'https://passport.ustc.edu.cn/login?service=https%3A%2F%2Fweixine.ustc.edu.cn%2F2020%2Fcaslogin',
	cookie_jar)

	# Get CAS token
	login_form_data = etree.HTML(r.text)
	token_line = login_form_data.xpath("//input[@id='CAS_LT']/@value")
	assert(len(token_line) == 1)
	r.close()
	return token_line[0]

def visit_daily_report_page(req, cookie_jar, token):
	global USERNAME, PASSWORD
	login_payload = {
		'username': USERNAME,
		'password': PASSWORD,
		'CAS_LT': token,
		'service': 'https://weixine.ustc.edu.cn/2020/caslogin',
		'button': '',
		'showCode': '',
		'warn': '',
		'model': 'uplogin.jsp'
	}

	dout(login_payload)
	# Login start
	r = visit(req, 'post',
	'https://passport.ustc.edu.cn/login',
	cookie_jar, login_payload)
	# dout(cookie_jar.keys())
	dout('CAS-Token: %s' % token)
	#dout(r.text)
	#return 0

	# Get my token for later commit
	login_form_data = etree.HTML(r.text)
	token_line = login_form_data.xpath("//*[@id='daliy-report']/form/input/@value")
	assert(len(token_line) == 1)
	# Close login request
	r.close()
	return token_line[0]

def commit_daily_report(req, cookie_jar, token):
	dout('=======================')
	# Prepare for report request
	headers = { 
        'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.61 Safari/537.36'}
	param = {
		# 'Accept': 'text/html, application/xhtml+xml, application/xml; q=0.9, image/webp,image/apng, */*; q=0.8, application/signed-exchange; v=b3; q=0.9',
		'Accept - Encoding': 'gzip, deflate, br',
		'Accept - Language': 'en-US,en;q=0.9,zh-CN;q=0.8,zh;q=0.7',
		'Cache - Control': 'max-age=0',
		'Content - Type': 'application/x-www-form-urlencoded',
		'Origin': 'https://weixine.ustc.edu.cn',
		'Referer': 'https://weixine.ustc.edu.cn/2020/home',
		'Src - Fetch - Dest': 'document',
		'Src - Fetch - Mode': 'navigate',
		'Src - Fetch - Site': 'same-origin',
		'Src - Fetch - User': '71',
		'Upgrade - Insecure - Requests': '1'
	}
	report_payload = {
		'_token': token,				# 加入上面获得的token
		#'now_address' : '1',			# 当前所在地：内地
		#'gps_now_address': '',			#
		#'now_province': NOW_PROVINCE,		# 当前所在地（省或直辖市）
		#'gps_province': '',				#
		#'now_city': NOW_CITY,			# 当前所在地（地级市）
		#'gps_city': '',					#
		#'now_country': NOW_DISTRICT,			# 当前所在地（区）
		#'gps_country': '',			#
		#'now_detail': '',				#
		#'is_inschool': WHICH_CAMPUS,				# 是否在校：选择对应校区
		'juzhudi': CAMPUS_NAME,		# 当前校区：（中文）
		'city_area':	HOME_DIST,  # 具体位置：区县选择
		'jutiwz':			HOME_ADDR,  # 具体位置：住址
		'body_condition':	'1',		# 当前身体状况：正常
		'body_condition_detail': '',	# 
		'now_status': NOW_STATUS,				# 当前状态：正常在校园内
		'now_status_detail': '',		#
		'has_fever': '0',				# 当前有无发热症状：无
		'last_touch_sars': '0',			# 有无接触患者：无
		'last_touch_sars_date': '',		#
		'last_touch_sars_detail': '',	#
		'is_danger': '0',					# 当日是否到过疫情中高风险地区：无
		'is_goto_danger': '0',			# 14天内是否有疫情中高风险地区旅居史：无
		'jinji_lxr': EMERGENCY_CONTACT,						# 紧急联系人
		'jinji_guanxi': EMERGENCY_CONTACT_RELATION,				# （紧急联系人）与本人关系
		'jiji_mobile': EMERGENCY_CONTACT_PHONE,	# 紧急联系人电话
		'other_detail': '',				# 其他情况说明：（无）
	}
	dout(report_payload)

	r = visit(req, 'post',
	'https://weixine.ustc.edu.cn/2020/daliy_report',
	cookie_jar, report_payload, headers, param)
	# dout(cookie_jar.items())

	dout('Last status code: %d' % r.status_code)
	#dout(r.text)
	if (r.status_code == 200):
		ret_text = r.text
		last_report_info_pos = r.text.find('上次上报时间')
		dout(ret_text[last_report_info_pos:(last_report_info_pos+26)])
		r.close()
		return 0
	else:
		sys.stderr.write('错误！无法完成上报。')
		sys.stderr.flush()
		return -1

def main():
	global args
	args = parse_args()

	#check_global_input()

	# Prepare for the session
	req = requests.Session()
	cookie_jar = RequestsCookieJar()

	# Random delay
	dout(args.no_sleep and '(Skip wait)' or 'Randomly delaying 1-300 seconds...')
	sleep(random.randint(1,300))

	token = visit_login_page(req, cookie_jar)
	token = visit_daily_report_page(req, cookie_jar, token)

	# Random delay before requesting daily report
	dout(args.no_sleep and '(Skip wait)' or 'Randomly delaying 10-20 seconds...')
	sleep(random.randint(10,20))

	return commit_daily_report(req, cookie_jar, token)

if __name__ == '__main__':
	sys.exit(main())


weekly-apply.py
#!/usr/bin/env python3

######################################################
# 自动打卡的参数，请在这改

## 必填信息：

# 科大统一认证账号密码
# （明文保存。请勿分享带密码的脚本给他人）
USERNAME = ''
PASSWORD = ''

# 后面的内容不需要修改。如不能用请提issue
######################################################
import sys
import requests
from requests.cookies import RequestsCookieJar
from lxml import etree

import time
import random

import datetime

import argparse

# Arguments from command line
args = None

dout = lambda line: (args.debug and print(line))
sleep = lambda n: (not args.no_sleep and time.sleep(n))

def parse_args():
	global args
	parser = argparse.ArgumentParser()
	parser.add_argument('-d', '--debug', action='store_true', help='Debug mode: show debug info')
	parser.add_argument('-i', '--no-sleep', action='store_true', help='Debug mode: no sleep')
	return parser.parse_args()

def check_global_input():
	# Check username and password
	global USERNAME, PASSWORD
	if not USERNAME:	USERNAME = input('请输入学号：')
	if not PASSWORD:	PASSWORD = input('请输入密码：')

def visit(req, fn_name, url, cookie_jar, payload=None, headers=None, param=None):
	fn_list = {'get': req.get, 'post': req.post}
	try:
		fn = fn_list[fn_name.lower()]
	except KeyError:
		raise ValueError('Unsupported visit method: %s' % fn_name)

	dout('Requesting %s' % url)
	r = fn(url, cookies=cookie_jar, data=payload, headers=headers, params=param,
		allow_redirects=False, timeout=50)
	# Trace redirections: keep all cookies
	while r.status_code in range(300, 304):
		url = r.headers['Location']
		dout('Redirecting to %s' % url)
		cookie_jar.update(r.cookies)
		r = req.get(url, cookies=cookie_jar, allow_redirects=False, timeout=50)
	
	# Finally update my cookies
	cookie_jar.update(r.cookies)
	return r

def visit_login_page(req, cookie_jar):
	# Get CAS login page
	r = visit(req, 'get',
	'https://passport.ustc.edu.cn/login?service=https%3A%2F%2Fweixine.ustc.edu.cn%2F2020%2Fcaslogin',
	cookie_jar)

	# Get CAS token
	login_form_data = etree.HTML(r.text)
	token_line = login_form_data.xpath("//input[@id='CAS_LT']/@value")
	assert(len(token_line) == 1)
	r.close()
	return token_line[0]

def visit_daily_report_page(req, cookie_jar, token):
	global USERNAME, PASSWORD
	login_payload = {
		'username': USERNAME,
		'password': PASSWORD,
		'CAS_LT': token,
		'service': 'https://weixine.ustc.edu.cn/2020/caslogin',
		'button': '',
		'showCode': '',
		'warn': '',
		'model': 'uplogin.jsp'
	}

	dout(login_payload)
	# Login start
	r = visit(req, 'post',
	'https://passport.ustc.edu.cn/login',
	cookie_jar, login_payload)
	# dout(cookie_jar.keys())
	dout('CAS-Token: %s' % token)
	#dout(r.text)
	#return 0

	# Get my token for later commit
	login_form_data = etree.HTML(r.text)
	token_line = login_form_data.xpath("//*[@id='daliy-report']/form/input/@value")
	assert(len(token_line) == 1)
	# Close login request
	r.close()
	return token_line[0]

def visit_weekly_apply_page(req, cookie_jar):
	# Get weekly apply page
	r = visit(req, 'get',
	'https://weixine.ustc.edu.cn/2020/apply/daliy',
	cookie_jar)

	# Get apply token
	apply_form_data = etree.HTML(r.text)
	#token_line = apply_form_data.xpath("//input[@name='_token']/@value")
	token_line = apply_form_data.xpath("//*[@id='daliy-report']/form/input/@value")
	assert(len(token_line) == 1)
	r.close()
	return token_line[0]

def commit_weekly_apply(req, cookie_jar, token):
	start = datetime.datetime.today()
	end = start + datetime.timedelta(days = 6)
	apply_payload = {
		'_token': token,
		'start_date': start.strftime('%Y-%m-%d'),
		'end_date': end.strftime('%Y-%m-%d')
	}
	dout(apply_payload)
	# Apply start
	r = visit(req, 'post',
	'https://weixine.ustc.edu.cn/2020/apply/daliy/post',
	cookie_jar, apply_payload)
	dout('Last status code: %d' % r.status_code)
	if (r.status_code == 200):
		ret_text = r.text
		last_apply_info_pos = ret_text.find('报备成功')
		dout(ret_text[last_apply_info_pos:(last_apply_info_pos+15)])
		r.close()
		return 0
	else:
		sys.stderr.write('错误！无法完成申请。')
		sys.stderr.flush()
		return -1

def main():
	global args
	args = parse_args()

	#check_global_input()

	# Prepare for the session
	req = requests.Session()
	cookie_jar = RequestsCookieJar()

	# Random delay
	dout(args.no_sleep and '(Skip wait)' or 'Randomly delaying 1-300 seconds...')
	sleep(random.randint(1,300))

	token = visit_login_page(req, cookie_jar)
	visit_daily_report_page(req, cookie_jar, token)

	# Random delay before requesting weekly apply page
	dout(args.no_sleep and '(Skip wait)' or 'Randomly delaying 10-20 seconds...')
	sleep(random.randint(10,20))

	token = visit_weekly_apply_page(req, cookie_jar)

	# Random delay before requesting weekly apply commit
	dout(args.no_sleep and '(Skip wait)' or 'Randomly delaying 10-20 seconds...')
	sleep(random.randint(10,20))
	
	return commit_weekly_apply(req, cookie_jar, token)

if __name__ == '__main__':
	sys.exit(main())


daily-report-out-campus.py
#!/usr/bin/env python3

######################################################
# 自动打卡的参数，请在这改

## 必填信息：

# 科大统一认证账号密码
# （明文保存。请勿分享带密码的脚本给他人）
USERNAME = ''
PASSWORD = ''

# 当前状态：
# 1: 正常在校园内
# 2: 正常在家
# 如非这两种情况请手动打卡
NOW_STATUS = '1'

# 2022/03/22更新：校外居住同学这里不需要再写了，因为打卡界面设置了不允许修改
# （虽然提交请求时硬改是可以的^_^）
CAMPUS_NAME = '合肥市内校外'

# 2022/03/30更新：校外居住报备时需要填写具体位置。校内同学可跳过
## 区县选择：只能选择以下四个区之一：包河区、蜀山区、庐阳区、瑶海区
HOME_DIST = ''
#HOME_DIST = '包河区'
HOME_ADDR = ''

# 紧急联系人：姓名、关系、电话
EMERGENCY_CONTACT = ''
EMERGENCY_CONTACT_RELATION = ''
EMERGENCY_CONTACT_PHONE = ''

# 2022/05/05更新：系统中增加进出校报备选项，不再需要向老师申请
# 2022/03/22更新：进出校申请需要选择老师审核！！这里的编号是老师在系统里的编号，
# 一般是10位数字，不知道请运行脚本check-advisor-id.py获取。
# 注意：不是老师的工号，不要乱写！！
#ADVISOR_ID = ''

# 2022/03/22更新：进出校原因。需要可自行更改。
GETINSCHOOL_REASON = '日常工作学习需要进出校园。'

# 2022/03/22更新：校外同学选择“进出校申请”时同样需要选择进入哪个校区，选择方法相同。
# 2022/03/20更新：必须选择“跨校区目的地”中的一项或多项！
#   所有可选项为：东/西/南/北/中/高新校区、先研院、国金院
# 可在下面提供的默认选项中选择一行去掉注释即可，或自行修改。多行时只有最后一行会生效。
# 注意：修改时请使用英文逗号和引号，在下面的花括号中手动添加或修改校区名
DESTINATION_CAMPUS = {'东校区', '西校区', '高新校区', '先研院'}
#DESTINATION_CAMPUS = {'东校区', '西校区'}
#DESTINATION_CAMPUS = {'高新校区', '先研院'}
#DESTINATION_CAMPUS = {} # 自行填入

# 后面的内容不需要修改。如不能用请提issue
######################################################
import sys
import requests
from requests.cookies import RequestsCookieJar
from lxml import etree

import time
import random

import datetime

import argparse

# Arguments from command line
args = None

dout = lambda line: (args.debug and print(line))
sleep = lambda n: (not args.no_sleep and time.sleep(n))

def parse_args():
	global args
	parser = argparse.ArgumentParser()
	parser.add_argument('-d', '--debug', action='store_true', help='Debug mode: show debug info')
	parser.add_argument('-i', '--no-sleep', action='store_true', help='Debug mode: no sleep')
	return parser.parse_args()

def check_global_input():
	# Check username and password
	global USERNAME, PASSWORD
	if not USERNAME:	USERNAME = input('请输入学号：')
	if not PASSWORD:	PASSWORD = input('请输入密码：')

def visit(req, fn_name, url, cookie_jar, payload=None, headers=None, param=None):
	fn_list = {'get': req.get, 'post': req.post}
	try:
		fn = fn_list[fn_name.lower()]
	except KeyError:
		raise ValueError('Unsupported visit method: %s' % fn_name)

	dout('Requesting %s' % url)
	r = fn(url, cookies=cookie_jar, data=payload, headers=headers, params=param,
		allow_redirects=False, timeout=50)
	# Trace redirections: keep all cookies
	while r.status_code in range(300, 304):
		url = r.headers['Location']
		dout('Redirecting to %s' % url)
		cookie_jar.update(r.cookies)
		r = req.get(url, cookies=cookie_jar, allow_redirects=False, timeout=50)
	
	# Finally update my cookies
	cookie_jar.update(r.cookies)
	return r

def visit_login_page(req, cookie_jar):
	# Get CAS login page
	r = visit(req, 'get',
	'https://passport.ustc.edu.cn/login?service=https%3A%2F%2Fweixine.ustc.edu.cn%2F2020%2Fcaslogin',
	cookie_jar)

	# Get CAS token
	login_form_data = etree.HTML(r.text)
	token_line = login_form_data.xpath("//input[@id='CAS_LT']/@value")
	assert(len(token_line) == 1)
	r.close()
	return token_line[0]

def visit_daily_report_page(req, cookie_jar, token):
	global USERNAME, PASSWORD
	login_payload = {
		'username': USERNAME,
		'password': PASSWORD,
		'CAS_LT': token,
		'service': 'https://weixine.ustc.edu.cn/2020/caslogin',
		'button': '',
		'showCode': '',
		'warn': '',
		'model': 'uplogin.jsp'
	}

	dout(login_payload)
	# Login start
	r = visit(req, 'post',
	'https://passport.ustc.edu.cn/login',
	cookie_jar, login_payload)
	# dout(cookie_jar.keys())
	dout('CAS-Token: %s' % token)
	#dout(r.text)
	#return 0

	# Get my token for later commit
	login_form_data = etree.HTML(r.text)
	token_line = login_form_data.xpath("//*[@id='daliy-report']/form/input/@value")
	assert(len(token_line) == 1)
	# Close login request
	r.close()
	return token_line[0]

def commit_daily_report(req, cookie_jar, token):
	dout('=======================')
	# Prepare for report request
	headers = { 
		'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.61 Safari/537.36'
	}
	param = {
		# 'Accept': 'text/html, application/xhtml+xml, application/xml; q=0.9, image/webp,image/apng, */*; q=0.8, application/signed-exchange; v=b3; q=0.9',
		'Accept - Encoding': 'gzip, deflate, br',
		'Accept - Language': 'en-US,en;q=0.9,zh-CN;q=0.8,zh;q=0.7',
		'Cache - Control': 'max-age=0',
		'Content - Type': 'application/x-www-form-urlencoded',
		'Origin': 'https://weixine.ustc.edu.cn',
		'Referer': 'https://weixine.ustc.edu.cn/2020/home',
		'Src - Fetch - Dest': 'document',
		'Src - Fetch - Mode': 'navigate',
		'Src - Fetch - Site': 'same-origin',
		'Src - Fetch - User': '71',
		'Upgrade - Insecure - Requests': '1'
	}
	report_payload = {
		'_token': token,				# 加入上面获得的token
		#'now_address' : '1',			# 当前所在地：内地
		#'gps_now_address': '',			#
		#'now_province': NOW_PROVINCE,		# 当前所在地（省或直辖市）
		#'gps_province': '',				#
		#'now_city': NOW_CITY,			# 当前所在地（地级市）
		#'gps_city': '',					#
		#'now_country': NOW_DISTRICT,			# 当前所在地（区）
		#'gps_country': '',			#
		#'now_detail': '',				#
		#'is_inschool': WHICH_CAMPUS,				# 是否在校：选择对应校区
		'juzhudi': CAMPUS_NAME,		# 当前校区：（中文）
		'city_area':	HOME_DIST,  # 具体位置：区县选择
		'jutiwz':			HOME_ADDR,  # 具体位置：住址
		'body_condition':	'1',		# 当前身体状况：正常
		'body_condition_detail': '',	# 
		'now_status': NOW_STATUS,				# 当前状态：正常在校园内
		'now_status_detail': '',		#
		'has_fever': '0',				# 当前有无发热症状：无
		'last_touch_sars': '0',			# 有无接触患者：无
		'last_touch_sars_date': '',		#
		'last_touch_sars_detail': '',	#
		'is_danger': '0',					# 当日是否到过疫情中高风险地区：无
		'is_goto_danger': '0',			# 14天内是否有疫情中高风险地区旅居史：无
		'jinji_lxr': EMERGENCY_CONTACT,						# 紧急联系人
		'jinji_guanxi': EMERGENCY_CONTACT_RELATION,				# （紧急联系人）与本人关系
		'jiji_mobile': EMERGENCY_CONTACT_PHONE,	# 紧急联系人电话
		'other_detail': '',				# 其他情况说明：（无）
	}
	dout(report_payload)

	r = visit(req, 'post',
	'https://weixine.ustc.edu.cn/2020/daliy_report',
	cookie_jar, report_payload, headers, param)
	# dout(cookie_jar.items())

	dout('Last status code: %d' % r.status_code)
	#dout(r.text)
	if (r.status_code == 200):
		ret_text = r.text
		last_report_info_pos = r.text.find('上次上报时间')
		dout(ret_text[last_report_info_pos:(last_report_info_pos+26)])
		r.close()
		return 0
	else:
		sys.stderr.write('错误！无法完成上报。')
		sys.stderr.flush()
		return -1

def visit_get_into_campus_application_page(req, cookie_jar):
	# Get application page
	r = visit(req, 'get',
	'https://weixine.ustc.edu.cn/2020/apply/daliy/i?t=3',
	cookie_jar)

	# Get my token for later commit
	login_form_data = etree.HTML(r.text)
	token_line = login_form_data.xpath("//*[@id='daliy-report']/form/input/@value")
	assert(len(token_line) == 1)
	# Close application request
	r.close()
	return token_line[0]

def commit_get_into_campus_application(req, cookie_jar, token):
	dout('=======================')
	# Prepare for application request
	headers = { 
		'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.61 Safari/537.36'
	}
	param = {
		# 'Accept': 'text/html, application/xhtml+xml, application/xml; q=0.9, image/webp,image/apng, */*; q=0.8, application/signed-exchange; v=b3; q=0.9',
		'Accept - Encoding': 'gzip, deflate, br',
		'Accept - Language': 'en-US,en;q=0.9,zh-CN;q=0.8,zh;q=0.7',
		'Cache - Control': 'max-age=0',
		'Content - Type': 'application/x-www-form-urlencoded',
		'Origin': 'https://weixine.ustc.edu.cn',
		'Referer': 'https://weixine.ustc.edu.cn/2020/stayinout_apply?t=3',
		'Src - Fetch - Dest': 'document',
		'Src - Fetch - Mode': 'navigate',
		'Src - Fetch - Site': 'same-origin',
		'Src - Fetch - User': '71',
		'Upgrade - Insecure - Requests': '1'
	}

	# 这里不要尝试改时间了，已经试过了没用
	start = datetime.datetime.today()
	#tomorrow = start + datetime.timedelta(days = 1)
	end = datetime.datetime(
		start.year,
		start.month,
		start.day,
		hour=23,
		minute=59,
		second=59
	)
	payload = {
		'_token': token,
		'start_date': start.strftime('%Y-%m-%d %H:%M:%S'),
		'end_date': end.strftime('%Y-%m-%d %H:%M:%S'),
		'reason': GETINSCHOOL_REASON,
		'return_college[]': DESTINATION_CAMPUS,
		't': 3	# 进出校申请
	}
	dout(payload)
	# Application start
	r = visit(req, 'post',
	'https://weixine.ustc.edu.cn/2020/apply/daliy/post',
	cookie_jar, payload)
	dout('Last status code: %d' % r.status_code)
	if (r.status_code == 200):
		result_page = etree.HTML(r.text)
		applications_box_headers = result_page.xpath("//div[@id='apply-box']/table/thead/tr/th/text()")
		dout('\t'.join(applications_box_headers[:3]))
		applications_box_entries = result_page.xpath("//div[@id='apply-box']/table/tbody/tr")
		for i in range(1, len(applications_box_entries) + 1):
			applications_box_entry_period_start = result_page.xpath("//div[@id='apply-box']/table/tbody/tr[%d]/td[1]/text()" % i)[0]
			applications_box_entry_period_end = result_page.xpath("//div[@id='apply-box']/table/tbody/tr[%d]/td[2]/text()" % i)[0]
			applications_box_entry_status = result_page.xpath("//div[@id='apply-box']/table/tbody/tr[%d]/td[3]/text()" % i)[0]
			dout(f'{applications_box_entry_period_start}\t{applications_box_entry_period_end}\t{applications_box_entry_status}')
		r.close()
		return 0
	else:
		sys.stderr.write('错误！无法完成申请。')
		sys.stderr.flush()
		return -1

def main():
	global args
	args = parse_args()

	#check_global_input()

	# Prepare for the session
	req = requests.Session()
	cookie_jar = RequestsCookieJar()

	# Random delay
	dout(args.no_sleep and '(Skip wait)' or 'Randomly delaying 1-300 seconds...')
	sleep(random.randint(1,300))

	token = visit_login_page(req, cookie_jar)
	token = visit_daily_report_page(req, cookie_jar, token)

	# Random delay before requesting daily report
	dout(args.no_sleep and '(Skip wait)' or 'Randomly delaying 10-20 seconds...')
	sleep(random.randint(10,20))

	# Committing daily report
	if commit_daily_report(req, cookie_jar, token) < 0:
		return -1

	# Random delay before visiting outside-campus application page
	dout(args.no_sleep and '(Skip wait)' or 'Randomly delaying 10-20 seconds...')
	sleep(random.randint(10,20))

	# Visiting outside-campus application page: behave more like a real person
	# Actually this is useless
	visit(req, 'get',
	'https://weixine.ustc.edu.cn/2020/apply/daliy',
	cookie_jar)

	# Random delay before committing cross-campus application
	dout(args.no_sleep and '(Skip wait)' or 'Randomly delaying 10-20 seconds...')
	sleep(random.randint(10,20))

	token = visit_get_into_campus_application_page(req, cookie_jar)

	# Random delay before committing cross-campus application
	dout(args.no_sleep and '(Skip wait)' or 'Randomly delaying 10-20 seconds...')
	sleep(random.randint(10,20))

	return commit_get_into_campus_application(req, cookie_jar, token)

if __name__ == '__main__':
	sys.exit(main())
